File: src/main/java/cspom/CSPOM.java
/**
 *  CSP Object Model (CSPOM) - a CSP modeling API
 *  Copyright (C) 2008-2010 Julien Vion
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.

 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, 
 *  MA 02110-1301  USA
 */
package cspom;

import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.zip.GZIPInputStream;

import org.apache.tools.bzip2.CBZip2InputStream;

import cspom.compiler.ConstraintCompiler;
import cspom.compiler.PredicateParseException;
import cspom.constraint.CSPOMConstraint;
import cspom.constraint.FunctionalConstraint;
import cspom.variable.BooleanDomain;
import cspom.variable.CSPOMDomain;
import cspom.variable.CSPOMVariable;
import cspom.xcsp.Parser;
import cspom.xcsp.XCSPParseException;

/**
 * 
 * CSPOM is the central class of the Constraint Satisfaction Problem Object
 * Model. You can create a problem from scratch by instantiating this class and
 * then using the ctr and var methods. The CSPOM.load() method can be used in
 * order to create a CSPOM object from an XCSP file.
 * <p>
 * The CSPOM class adheres to the following definition :
 * 
 * A CSP is defined as a pair (X, C), X being a finite set of variables, and C a
 * finite set of constraints.
 * 
 * A domain is associated to each variable x in X. Each constraint involves a
 * finite set of variables (its scope) and defines a set of allowed and
 * forbidden instantiations of these variables.
 * 
 * @author Julien Vion
 * @see CSPOMConstraint
 * @see CSPOMVariable
 * 
 */
public final class CSPOM {

	/**
	 * Version obtained from SVN Info.
	 */
	public static final String VERSION;

	static {
		final Matcher matcher = Pattern.compile("Rev:\\ (\\d+)").matcher(
				"$Rev$");
		matcher.find();
		VERSION = matcher.group(1);
	}

	/**
	 * List of variables of the problem, in order.
	 */
	private final List<CSPOMVariable> variableList;

	/**
	 * Map used to easily retrieve a variable according to its name.
	 */
	private final Map<String, CSPOMVariable> variableMap;

	/**
	 * Collection of all constraints of the problem.
	 */
	private final Collection<CSPOMConstraint> constraints;

	/**
	 * The constraint compiler used by this CSPOM instance.
	 */
	private ConstraintCompiler constraintCompiler;

	/**
	 * Creates an empty problem, without any initial variables nor constraints.
	 */
	public CSPOM() {
		variableList = new LinkedList<CSPOMVariable>();
		variableMap = new HashMap<String, CSPOMVariable>();
		constraints = new ArrayList<CSPOMConstraint>();
	}

	/**
	 * Opens an InputStream according to the given URL. If URL ends with ".gz"
	 * or ".bz2", the stream is inflated accordingly.
	 * 
	 * @param url
	 *            URL to open
	 * @return An InputStream corresponding to the given URL
	 * @throws IOException
	 *             If the InputStream could not be opened
	 */
	public static InputStream problemInputStream(final URL url)
			throws IOException {

		final String path = url.getPath();

		if (path.endsWith(".gz")) {
			return new GZIPInputStream(url.openStream());
		}

		if (path.endsWith(".bz2")) {
			final InputStream is = url.openStream();
			is.read();
			is.read();
			return new CBZip2InputStream(is);
		}

		return url.openStream();
	}

	/**
	 * Loads a CSPOM from a given XCSP file.
	 * 
	 * @param xcspFile
	 *            Either a filename or an URI. Filenames ending with .gz or .bz2
	 *            will be inflated accordingly.
	 * @return The loaded CSPOM
	 * @throws XCSPParseException
	 *             If the given file could not be parsed.
	 * @throws IOException
	 *             If the given file could not be read.
	 */
	public static CSPOM load(final String xcspFile) throws XCSPParseException,
			IOException {
		final URI uri;
		try {
			uri = new URI(xcspFile);
		} catch (URISyntaxException e) {
			throw new IOException("Invalid URI", e);
		}

		if (uri.getScheme() == null) {
			return load(new URL("file://" + uri));
		}

		return load(uri.toURL());
	}

	/**
	 * Loads a CSPOM from a given XCSP file.
	 * 
	 * @param url
	 *            An URL locating the XCSP file. Filenames ending with .gz or
	 *            .bz2 will be inflated accordingly.
	 * @return The loaded CSPOM
	 * @throws XCSPParseException
	 *             If the given file could not be parsed.
	 * @throws IOException
	 *             If the given file could not be read.
	 */
	public static CSPOM load(final URL url) throws XCSPParseException,
			IOException {

		final CSPOM problem = new CSPOM();
		final InputStream problemIS = problemInputStream(url);

		new Parser(problem).parse(problemIS);
		// problem.compile();
		return problem;
	}

	public void compile() {
		final Collection<CSPOMConstraint> toRemove = new ArrayList<CSPOMConstraint>();

		for (CSPOMVariable v : getVariables()) {
			if (BooleanDomain.TRUE.equals(v.getDomain())) {
				for (CSPOMConstraint c : v.getConstraints()) {
					if (!(c instanceof FunctionalConstraint)
							|| !"and".equals(c.getDescription())) {
						continue;
					}
					final FunctionalConstraint fc = (FunctionalConstraint) c;
					if (fc.getResultVariable() == v) {
						for (CSPOMVariable a : fc.getArguments()) {
							a.setDomain(BooleanDomain.TRUE);
						}
						toRemove.add(c);
					}
				}
			}
		}

		for (CSPOMConstraint c : toRemove) {
			removeConstraint(c);
		}

	}

	private void removeConstraint(final CSPOMConstraint c) {
		for (CSPOMVariable v : c.getScope()) {
			v.removeConstraint(c);
		}
		constraints.remove(c);

	}

	public void removeVariable(final CSPOMVariable v) {
		if (v.getConstraints().isEmpty()) {
			variableList.remove(v);
			variableMap.remove(v.getName());
		} else {
			throw new IllegalArgumentException(v
					+ " is still implied by constraints");
		}

	}

	/**
	 * @return The variables of this problem.
	 */
	public List<CSPOMVariable> getVariables() {
		return variableList;
	}

	/**
	 * Adds a variable to the problem.
	 * 
	 * @param variable
	 *            The variable to add. It must have an unique name.
	 * @throws DuplicateVariableException
	 *             If a variable with the same name already exists.
	 */
	public void addVariable(final CSPOMVariable variable)
			throws DuplicateVariableException {
		if (variableMap.put(variable.getName(), variable) != null) {
			throw new DuplicateVariableException(variable.getName());
		}
		variableList.add(variable);

	}

	/**
	 * Adds a constraint to the problem.
	 * 
	 * @param constraint
	 *            The constraint to add.
	 */
	public void addConstraint(final CSPOMConstraint constraint) {
		constraints.add(constraint);
	}

	/**
	 * @return The constraints of this problem.
	 */
	public Collection<CSPOMConstraint> getConstraints() {
		return constraints;
	}

	/**
	 * Adds a bounded, unnamed variable in the problem.
	 * 
	 * @param lb
	 *            lower bound
	 * @param ub
	 *            upper bound
	 * @return The added variable.
	 */
	public CSPOMVariable var(final int lb, final int ub) {
		final CSPOMVariable variable = new CSPOMVariable(lb, ub);
		try {
			addVariable(variable);
		} catch (DuplicateVariableException e) {
			throw new IllegalStateException();
		}
		return variable;
	}

	/**
	 * Adds a bounded, named variable in the problem.
	 * 
	 * @param name
	 *            name of the variable
	 * @param lb
	 *            lower bound
	 * @param ub
	 *            upper bound
	 * @return The added variable.
	 * @throws DuplicateVariableException
	 *             if a variable of the same name already exists
	 */
	public CSPOMVariable var(final String name, final int lb, final int ub)
			throws DuplicateVariableException {
		final CSPOMVariable variable = new CSPOMVariable(name, lb, ub);
		addVariable(variable);
		return variable;
	}

	/**
	 * Adds functional constraints to the problem. The given predicate will be
	 * parsed and the appropriate constraints added to the problem. The
	 * predicate may be complex and is usually translated to a set of
	 * constraints and auxiliary variables, resulting in the returned
	 * subproblem. Use variable names in the predicate to reference them. These
	 * variables must already be added to the problem.
	 * 
	 * @param string
	 *            A predicate.
	 * @throws PredicateParseException
	 *             If an error was encountered parsing the predicate.
	 * @return The subproblem corresponding to the given predicate
	 */
	public CSPOM ctr(final String string) throws PredicateParseException {
		if (constraintCompiler == null) {
			constraintCompiler = new ConstraintCompiler(this);
		}
		final CSPOM subproblem = constraintCompiler.split(string);
		merge(subproblem);
		return subproblem;
	}

	/**
	 * @param variableName
	 *            A variable name.
	 * @return The variable with the corresponding name.
	 */
	public CSPOMVariable getVariable(final String variableName) {
		return variableMap.get(variableName);
	}

	/**
	 * Merges the given subproblem in the current problem. If not already
	 * present, variables are added to the current problem.
	 * 
	 * @param problem
	 *            A subproblem.
	 */
	public void merge(final CSPOM problem) {
		for (CSPOMVariable v : problem.variableList) {
			if (!variableMap.containsValue(v)) {
				try {
					addVariable(v);
				} catch (DuplicateVariableException e) {
					throw new IllegalStateException();
				}
			}
		}

		constraints.addAll(problem.constraints);
	}

	@Override
	public String toString() {
		final StringBuilder stb = new StringBuilder();
		for (CSPOMVariable v : variableList) {
			stb.append(v).append(" = ");
			final CSPOMDomain<?> d = v.getDomain();
			if (d == null) {
				stb.append('?');
			} else {
				stb.append(d);
			}
			stb.append('\n');
		}
		for (CSPOMConstraint c : constraints) {
			stb.append(c).append('\n');
		}
		return stb.toString();
	}

	/**
	 * Generates the constraint network graph in the GML format. N-ary
	 * constraints are represented as nodes.
	 * 
	 * @return a String containing the GML representation of the constraint
	 *         network.
	 */
	public String toGML() {
		final StringBuilder stb = new StringBuilder();
		stb.append("graph [\n");
		stb.append("directed 0\n");
		for (CSPOMVariable v : variableList) {
			stb.append("node [\n");
			stb.append("id \"").append(v.getName()).append("\"\n");
			stb.append("label \"").append(v.getName()).append("\"\n");
			stb.append("]\n");
		}

		int gen = 0;
		for (CSPOMConstraint c : constraints) {
			if (c.getArity() > 2) {
				stb.append("node [\n");
				stb.append("id \"cons").append(gen).append("\"\n");
				stb.append("label \"").append(c.getDescription())
						.append("\"\n");
				stb.append("graphics [\n");
				stb.append("fill \"#FFAA00\"\n");
				stb.append("]\n");
				stb.append("]\n");

				for (CSPOMVariable v : c.getScope()) {
					stb.append("edge [\n");
					stb.append("source \"cons").append(gen).append("\"\n");
					stb.append("target \"").append(v.getName()).append("\"\n");
					stb.append("]\n");
				}
				gen++;
			} else if (c.getArity() == 2) {
				stb.append("edge [\n");
				stb.append("source \"").append(c.getScope().get(0)).append(
						"\"\n");
				stb.append("target \"").append(c.getScope().get(1)).append(
						"\"\n");
				stb.append("label \"").append(c.getDescription())
						.append("\"\n");
				stb.append("]\n");
			}
		}
		stb.append("]\n");

		return stb.toString();
	}
}

File: src/main/scala/cspom/CSPOM.scala
package cspom

import java.io.{IOException, InputStream}
import java.net.{URI, URL}

import com.typesafe.scalalogging.LazyLogging
import cspom.dimacs.CNFParser
import cspom.extension.{MDDRelation, Relation}
import cspom.flatzinc.FlatZincFastParser
import cspom.variable._
import cspom.xcsp.XCSP3Parser
import org.apache.commons.compress.compressors.{CompressorException, CompressorStreamFactory}

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer
import scala.language.implicitConversions
import scala.reflect.runtime.universe.TypeTag
import scala.util.{Failure, Try}


/**
  *
  * CSPOM is the central class of the Constraint Satisfaction Problem Object
  * Model. You can create a problem from scratch by instantiating this class and
  * then using the ctr and var methods. The CSPOM.load() method can be used in
  * order to create a CSPOM object from an XCSP file.
  * <p>
  * The CSPOM class adheres to the following definition :
  *
  * A CSP is defined as a pair (X, C), X being a finite set of variables, and C a
  * finite set of constraints.
  *
  * A domain is associated to each variable x in X. Each constraint involves a
  * finite set of variables (its scope) and defines a set of allowed and
  * forbidden instantiations of these variables.
  *
  * @author Julien Vion
  * @see CSPOMConstraint
  * @see CSPOMVariable
  *
  */
class CSPOM extends LazyLogging {

  val expressionMap = new ExpressionMap
  // private val ConstraintOrdering = Ordering.by { c: CSPOMConstraint[_] => c.id }
  private val ctrV = new mutable.LinkedHashMap[CSPOMExpression[_], mutable.LinkedHashSet[CSPOMConstraint[_]]]()

  //      .withDefaultValue(SortedSet.empty(ConstraintOrdering))
  //  private def getContainers(e: CSPOMExpression[_]): Option[Seq[(CSPOMSeq[_], Int)]] =
  //    Option(containers.get(e)).map(_.asScala.toSeq)
  private val annotations = mutable.HashMap[String, Annotations]().withDefaultValue(Annotations())
  /**
    * Collection of all constraints of the problem.
    */
  private val _constraints = mutable.LinkedHashMap[String, mutable.LinkedHashSet[CSPOMConstraint[_]]]()
  private var postponed: List[CSPOMConstraint[_]] = Nil
  private var _goal: Option[WithParam[CSPOMGoal[_]]] = None


  def setGoal(g: WithParam[CSPOMGoal[_]]): Unit = {
    // println(s"setGoal $g ${Thread.currentThread.getStackTrace.mkString(", ")}")
    resolvePostponed(g.obj.expr)
    this._goal = Some(g)
  }

  def goal: Option[WithParam[CSPOMGoal[_]]] = _goal

  def setGoal(g: CSPOMGoal[_], params: Map[String, Any] = Map()): Unit = setGoal(WithParam(g, params))


  def addAnnotation(expressionName: String, annotationName: String, annotation: Any): Unit = {
    annotations(expressionName) += (annotationName -> annotation)
  }

  def getAnnotations(expressionName: String): Annotations = annotations(expressionName)


  //def constraintSet: collection.Set[CSPOMConstraint[_]] = _constraints //.toSet

  def nameExpression[A <: CSPOMExpression[_]](e: A, n: String): A = {
    expressionMap.nameExpression(e, n)

    @annotation.tailrec
    def resolve(e: A): A = {
      postponed.find(_.flattenedScope.contains(e)) match {
        case Some(c) =>
          resolvePostponed(c)
          resolve(e)

        case _ => e
      }
    }

    resolve(e)

  }

  def removeConstraint(c: CSPOMConstraint[_]): Unit = {
    require(getConstraints(c.function)(c), s"$c not in problem")
    _constraints(c.function) -= c

    //require((Iterator(c.result) ++ c.arguments).forall(ctrV(_)(c)))

    for (v <- c.fullScope; ref <- ctrV.get(v)) {
      // Note: ctrV.get(v) may be empty if v appears several times in the scope
      ref -= c
      if (ref.isEmpty) {
        ctrV -= v
        if (!expressionMap.isReferenced(v)) {
          expressionMap.removeContainer(v)
        }
      }
    }
  }

  def getConstraints(function: String): collection.Set[CSPOMConstraint[_]] = _constraints.getOrElse(function, Set())

  def isReferenced(e: CSPOMExpression[_]): Boolean = ctrV.get(e).exists(_.nonEmpty) || expressionMap.isReferenced(e)

  def constraints(v: CSPOMExpression[_]): Seq[CSPOMConstraint[_]] = {
    /**
      * Should be kept immutable. ConstraintCompiler.replace will concurrently iterate and change
      * contents.
      */
    ctrV.get(v).toSeq.flatten
  }

  def unsafeConstraints(v: CSPOMExpression[_]): Iterator[CSPOMConstraint[_]] = {
    ctrV.get(v).iterator.flatten
  }

  def deepConstraints(v: CSPOMExpression[_]): Iterable[CSPOMConstraint[_]] = {
    val buf = new ArrayBuffer[CSPOMConstraint[_]]()

    def crawl(v: CSPOMExpression[_]): Unit = {
      ctrV.get(v).foreach(buf ++= _)
      for (
        (container, _) <- expressionMap.getContainers(v)
      ) {
        crawl(container)
      }
    }

    crawl(v)
    buf
  }

  def replaceExpression[R: TypeTag, T <: R](which: CSPOMExpression[R], by: CSPOMExpression[T]): Seq[(CSPOMExpression[_], CSPOMExpression[_])] = {
    val replacements = expressionMap.replaceExpression(which, by)

    for ((which, by) <- replacements; g <- goal) {
      g match {
        case WithParam(CSPOMGoal.Minimize(`which`), p) =>
          setGoal(CSPOMGoal.Minimize(by.asInstanceOf[CSPOMExpression[Int]]), p)
        case WithParam(CSPOMGoal.Maximize(`which`), p) =>
          setGoal(CSPOMGoal.Maximize(by.asInstanceOf[CSPOMExpression[Int]]), p)
        case _ =>
      }
    }

    replacements

  }

  def ctr[A](c: CSPOMConstraint[A]): CSPOMConstraint[A] = {
    ctrNetwork(c)
    c
  }

  def ctrNetwork(c: CSPOMConstraint[_]): Seq[CSPOMConstraint[_]] = {
    if (getConstraints(c.function)(c)) {
      logger.warn(s"$c already belongs to the problem")
      Nil
    } else {
      postpone(c)
      resolvePostponed(c)
    }
  }

  def ctr(v: SimpleExpression[Boolean]): CSPOMConstraint[Boolean] = {
    ctr(CSPOMConstraint("eq")(v, CSPOMConstant(true)))
  }

  def ctr(function: String)(arguments: CSPOMExpression[_]*): CSPOMConstraint[Boolean] =
    ctr(CSPOMConstraint(function)(arguments: _*))

  def ctr[A](result: CSPOMExpression[A])(function: String)(arguments: CSPOMExpression[_]*): CSPOMConstraint[A] =
    ctr(CSPOMConstraint(result)(function)(arguments: _*))

  def defineInt(f: SimpleExpression[Int] => CSPOMConstraint[_]): SimpleExpression[Int] = {
    define(IntVariable.free())(f)
  }

  def defineFree(f: SimpleExpression[_] => CSPOMConstraint[_]): SimpleExpression[_] =
    define(new FreeVariable())(f)

  def defineBool(f: SimpleExpression[Boolean] => CSPOMConstraint[_]): SimpleExpression[Boolean] =
    define(new BoolVariable())(f)

  def define[R](f: => R)(g: R => CSPOMConstraint[_]): R = {
    val r = f
    postpone(g(r))
    r
  }

  def postpone[A](c: CSPOMConstraint[A]): CSPOMConstraint[A] = {
    postponed ::= c
    c
  }

  def getPostponed: Seq[CSPOMConstraint[_]] = postponed

  override def toString: String = {
    val vars = referencedExpressions.map(e => (expressionMap.displayName(e), e)).sortBy(_._1).map {
      case (name, variable) => s"$name: $variable"
    }.mkString("\n")

    val cons = constraints.map(_.toString(expressionMap.displayName)).mkString("\n")

    s"$vars\n$cons\n${expressionMap.count} named expressions, ${ctrV.size} first-level expressions and ${constraints.size} constraints"
  }

  def constraints: Iterator[CSPOMConstraint[_]] = _constraints.valuesIterator.flatten

  def referencedExpressions: Seq[CSPOMExpression[_]] = {
    val involvedInConstraints = for (e <- ctrV.keysIterator; f <- e.flatten) yield f

    (involvedInConstraints ++ expressionMap.expressions).toSeq.distinct
  }

  def displayName(e: CSPOMExpression[_]): String = expressionMap.displayName(e)

  def namesOf(e: CSPOMExpression[_]): Iterable[String] = expressionMap.namesOf(e)

  def expression(name: String): Option[CSPOMExpression[_]] = expressionMap.expression(name)

  def variable(name: String): Option[CSPOMVariable[_]] = expressionMap.variable(name)

  def hasConstraint(c: CSPOMConstraint[_]): Boolean = _constraints(c.function)(c)

  /**
    * Adds a constraint to the problem.
    *
    * @param constraint
    * The constraint to add.
    */
  private[cspom] def addConstraint[A](constraint: CSPOMConstraint[A]): CSPOMConstraint[A] = {

    require(!getConstraints(constraint.function)(constraint),
      "The constraint " + constraint + " already belongs to the problem")

    _constraints.getOrElseUpdate(constraint.function, new mutable.LinkedHashSet()) += constraint

    for (
      v <- constraint.fullScope
    ) {
      ctrV.getOrElseUpdate(v, new mutable.LinkedHashSet()) += constraint
      expressionMap.registerContainer(v)
    }

    constraint
  }

  private def resolvePostponed(c: CSPOMConstraint[_]): Seq[CSPOMConstraint[_]] = {
    resolvePostponed(c.flattenedScope)
  }

  private def resolvePostponed(e: Iterable[CSPOMExpression[_]]): Seq[CSPOMConstraint[_]] = {
    val (posted, remaining) = resolvePostponed(e.toSet, postponed, Nil)
    postponed = remaining
    posted
  }

  @annotation.tailrec
  private def resolvePostponed(nodes: Set[CSPOMExpression[_]], postponed: List[CSPOMConstraint[_]], posted: List[CSPOMConstraint[_]]): (List[CSPOMConstraint[_]], List[CSPOMConstraint[_]]) = {
    val (post, remaining) = postponed.partition { c => c.flattenedScope.exists(nodes) }
    if (post.isEmpty) {
      (posted, remaining)
    } else {
      post.foreach(c => addConstraint(c))
      resolvePostponed(post.map(_.flattenedScope).reduce(_ ++ _), remaining, post ::: posted)
    }
  }

}

object CSPOM extends LazyLogging {

  type Parser = InputStream => Try[CSPOM]

  /**
    * Loads a CSPOM from a given XCSP3 file.
    *
    * @param file
    * Either a filename or an URI. Filenames ending with .gz or .bz2
    * will be inflated accordingly.
    * @return The loaded CSPOM
    */
  def loadXCSP3(file: String): Try[CSPOM] = load(file2url(file), XCSP3Parser)

  def file2url(file: String): URL = {
    val uri = new URI(file.replace(" ", "%20"))
    if (uri.isAbsolute && Option(uri.getScheme).isDefined) {
      uri.toURL
    } else {
      new URL("file:" + uri)
    }
  }

  /**
    * Loads a CSPOM from a given XCSP file.
    *
    * @param url
    * An URL locating the XCSP file. Filenames ending with .gz or
    * .bz2 will be inflated accordingly.
    * @return The loaded CSPOM and the list of original variable names
    */
  def load(url: URL, format: Parser): Try[CSPOM] = {
    problemInputStream(url).flatMap { pis =>
      val cspom = format(pis)
      pis.close()
      cspom
    }
  }

  /**
    * Opens an InputStream according to the given URL. If URL ends with ".gz"
    * or ".bz2", the stream is inflated accordingly.
    *
    * @param url
    * URL to open
    * @return An InputStream corresponding to the given URL
    * @throws IOException
    * If the InputStream could not be opened
    */
  @throws(classOf[IOException])
  def problemInputStream(url: URL): Try[InputStream] = Try {
    val is = url.openStream

    try {
      new CompressorStreamFactory().createCompressorInputStream(is)
    } catch {
      case e: CompressorException if e.getMessage == "No Compressor found for the stream signature." =>
        logger.info(e.getMessage)
        is
    }
  }

  def loadFZ(file: String): Try[CSPOM] = load(file2url(file), FlatZincFastParser)

  def loadCNF(file: String): Try[CSPOM] = load(file2url(file), CNFParser)

  def load(file: String): Try[CSPOM] = load(file2url(file))

  def load(url: URL): Try[CSPOM] =
    autoParser(url)
      .map(p => load(url, p))
      .getOrElse(Failure(new IllegalArgumentException(s"${url.getFile}: unknown file format")))

  def autoParser(url: URL): Option[Parser] = url.getFile match {
    case name if name.contains(".xml") => Some(XCSP3Parser)
    case name if name.contains(".cnf") => Some(CNFParser)
    case name if name.contains(".fzn") => Some(FlatZincFastParser)
    case _ => None
  }

  def apply(f: CSPOM => Any): CSPOM = {
    val p = new CSPOM()
    f(p)
    p
  }

  def ctr(v: SimpleExpression[Boolean])(implicit problem: CSPOM): CSPOMConstraint[Boolean] = problem.ctr(v)

  def ctr[A](c: CSPOMConstraint[A])(implicit problem: CSPOM): CSPOMConstraint[A] = problem.ctr(c)

  implicit class SeqOperations[A](vars: Seq[SimpleExpression[A]]) {
    def in(rel: Relation[A]): CSPOMConstraint[Boolean] =
      CSPOMConstraint("extension")(vars: _*) withParam("init" -> false, "relation" -> rel)

    def notIn(rel: Relation[A]): CSPOMConstraint[Boolean] = CSPOMConstraint("extension")(vars: _*) withParam("init" -> true, "relation" -> rel)
  }

  implicit class IntSeqOperations(vars: Seq[SimpleExpression[Int]]) {
    def in(rel: MDDRelation): CSPOMConstraint[Boolean] = SeqOperations(vars).in(rel)

    def notIn(rel: MDDRelation): CSPOMConstraint[Boolean] = SeqOperations(vars).notIn(rel)
  }

  //implicit def seq2Rel(s: Seq[Seq[Int]]): Relation[Int] = new Table(s)

  implicit def constant[A: TypeTag](c: A): CSPOMConstant[A] = CSPOMConstant(c)

  implicit def seq2CSPOMSeq[A: TypeTag](c: Iterable[CSPOMExpression[A]]): CSPOMSeq[A] = {
    val seq = c.toIndexedSeq
    CSPOMSeq(seq, seq.indices)
  }

  implicit def constantSeq[A <: AnyVal : TypeTag](c: Seq[A]): CSPOMSeq[A] = CSPOMSeq(c.map(constant): _*)

  implicit def matrix(sc: StringContext): Relation.MatrixContext = Relation.MatrixContext(sc)

  def goal(g: WithParam[CSPOMGoal[_]])(implicit problem: CSPOM): Unit = {
    problem.setGoal(g)
  }

  def goal(g: CSPOMGoal[_])(implicit problem: CSPOM): Unit = {
    problem.setGoal(g)
  }


}


